### PR TITLE
feat: add stake-weighted signature validation

### LINKED ISSUE #2491: Implement stake-weighted signature validation in Wrapped Blocks CLI tool
### Summary

The current signature validation logic uses simple node counting where each node's signature has equal weight. This is not production-correct — validation should use **stake-weighted consensus** based on each node's staking weight for the relevant period, matching how the consensus node and mirror node validate signatures.

### Current Behavior

In [`SignatureValidation.java:149-150`](tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/SignatureValidation.java#L149-L150):

```java
final int totalNodes = addressBook.nodeAddress().size();
final int threshold = (totalNodes / 3) + 1;
```

Each node's signature counts equally (weight = 1). The threshold is `ceil(nodeCount / 3)` valid signatures regardless of how much stake each signing node carries. A node with 1% of total stake counts the same as a node with 20%.

### Expected Behavior

Validation should use **stake weights** from `NodeStakeUpdate` transactions (issued once per staking period, ~daily at midnight UTC). The threshold should be based on the **Strong Minority** rule: `>= ceil(totalStake / 3)`.

### Pseudocode

```
// --- Phase 1: Collect stake weights from NodeStakeUpdate transactions ---
// During sequential block processing, intercept NodeStakeUpdate transactions
// and store per-node stake weights keyed by effective timestamp.

on NodeStakeUpdate(txBody, consensusTimestamp):
    for each nodeStake in txBody.node_stake:
        stakeRegistry[consensusTimestamp][nodeStake.node_id] = nodeStake.stake
    // stake (field 5) is the consensus weight — already capped/processed by the network

// --- Phase 2: Stake-weighted signature validation ---

function validateSignatures(block, blockProof, blockTime):
    addressBook = getAddressBookForBlock(blockTime)
    nodeStakes  = getStakesForBlock(blockTime)  // most recent NodeStakeUpdate before blockTime

    // Calculate total stake and threshold
    totalStake = 0
    for each node in addressBook.nodeAddress:
        totalStake += nodeStakes[node.nodeId] ?? 0

    if totalStake == 0:
        // Fallback: no stake data available (early blocks before first NodeStakeUpdate)
        // Use equal weighting — each node counts as 1
        totalStake = addressBook.nodeAddress.size()
        for each node in addressBook.nodeAddress:
            nodeStakes[node.nodeId] = 1

    // Strong Minority threshold: >= ceil(totalStake / 3)
    // Integer arithmetic equivalent (matches consensus node Threshold.java):
    threshold = (totalStake / 3) + (1 if totalStake % 3 != 0 else 0)

    // Verify signatures and accumulate stake
    validatedStake = 0
    validatedNodes = Set()

    for each sig in blockProof.signatures:
        if sig.nodeId in validatedNodes:
            continue  // skip duplicate signers

        pubKey = addressBook.getPublicKey(sig.nodeId)
        if verifyRsaSha384(pubKey, expectedHash, sig.bytes):
            validatedNodes.add(sig.nodeId)
            validatedStake += nodeStakes[sig.nodeId] ?? 0

            if validatedStake >= threshold:
                return SUCCESS  // early exit — consensus reached

    if validatedStake < threshold:
        FAIL("Insufficient stake: {validatedStake}/{totalStake}, need {threshold}")
```

### Reference Implementations

**Mirror Node** — [`ConsensusValidatorImpl.java`](https://github.com/hiero-ledger/hiero-mirror-node/blob/main/importer/src/main/java/org/hiero/mirror/importer/downloader/ConsensusValidatorImpl.java):
- Uses `BigDecimal.valueOf(totalStake).multiply(consensusRatio).setScale(0, RoundingMode.CEILING)` where `consensusRatio = 1/3`
- Groups verified signatures by file hash, sums stake per group, checks if any group meets threshold
- Falls back to equal weight (each node = 1) when no stake data exists

**Consensus Node** — [`Threshold.java`](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/platform-sdk/base-utility/src/main/java/org/hiero/base/utility/Threshold.java):
- `STRONG_MINORITY`: `part >= ceil(whole / 3)` using integer arithmetic: `part >= (whole / 3) + ((whole % 3 == 0) ? 0 : 1)`

**Mirror Node Node Stake Lookup** — [`AddressBookServiceImpl.java`](https://github.com/hiero-ledger/hiero-mirror-node/blob/main/importer/src/main/java/org/hiero/mirror/importer/addressbook/AddressBookServiceImpl.java):
- Reads `NodeStakeUpdate` transaction bodies to populate per-node stake values
- Combines address book entries with their corresponding stake into `ConsensusNodeWrapper`

### Relevant Protobuf

[`NodeStakeUpdateTransactionBody`](https://github.com/hashgraph/hedera-protobufs/blob/main/services/node_stake_update.proto) contains `repeated NodeStake node_stake` where each entry has:
- `int64 node_id` (field 3) — node identifier
- `int64 stake` (field 5) — **consensus weight** for the staking period (capped/processed, not raw HBAR)

Reference for NodeStakeUpdate transactions on mainnet: https://hashscan.io/mainnet/transactions?type=nodestakeupdate

### Implementation Tasks

1. **Parse `NodeStakeUpdate` transactions** — In `AddressBookRegistry` (or a new registry), intercept `NodeStakeUpdate` transaction bodies during sequential block processing and store the per-node stake weights, keyed by staking period
2. **Provide stake-for-block lookup** — Given a block timestamp, return the applicable stake weights (from the most recent `NodeStakeUpdate` before that block)
3. **Update `SignatureValidation.java`** — Replace simple node counting with stake-weighted accumulation:
   - Look up per-node stake for the block's time period
   - Compute `threshold = ceil(totalStake / 3)`
   - Sum stake of nodes with verified signatures instead of counting them
   - Fall back to equal-weight (current behavior) if no stake data is available (e.g., very early blocks before first `NodeStakeUpdate`)
4. **Update wrapped block signature validation** — Apply the same stake-weighted logic wherever signatures are validated during the wrapped block building pipeline (`ToWrappedBlocksCommand`, `ValidateDownloadLive`, etc.)
5. **Update tests** — Add unit tests for stake-weighted threshold logic including edge cases (unequal weights, missing stake data fallback, boundary conditions)

### Files Affected

- `tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/SignatureValidation.java` — core change
- `tools-and-tests/tools/src/main/java/org/hiero/block/tools/days/model/AddressBookRegistry.java` — add stake tracking
- `tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/ValidateBlocksCommand.java` — integration
- `tools-and-tests/tools/src/main/java/org/hiero/block/tools/days/subcommands/LiveSequential.java` — live edge signature waiting
- `tools-and-tests/tools/src/test/java/org/hiero/block/tools/blocks/validation/SignatureValidationTest.java` — tests

### PR DESCRIPTION
**Pull request by rockysingh**
## Summary
- Add stake-weighted consensus to `SignatureValidation` using `NodeStakeUpdate` transactions discovered from the block stream
- When stake data is available, threshold is `ceil(totalStake / 3)` (Strong Minority). Falls back to equal-weight mode (`each node = weight 1`, `threshold = nodeCount/3 + 1`) for pre-staking era blocks (before ~July 2022)
- New `NodeStakeRegistry` + `NodeStakeUpdateValidation` follow the existing `AddressBookRegistry` + `AddressBookUpdateValidation` pattern exactly

## Changes
| Action | File |
|--------|------|
| Create | `protobuf-sources/src/main/proto/internal/node_stake_history.proto` |
| Create | `NodeStakeRegistry.java` — thread-safe registry of stake snapshots with time-based lookup and JSON persistence |
| Create | `NodeStakeUpdateValidation.java` — sequential validation discovering `NodeStakeUpdate` txns from blocks |
| Modify | `SignatureValidation.java` — stake-weighted threshold with equal-weight fallback; backward-compatible constructor |
| Modify | `ValidateBlocksCommand.java` — wire `NodeStakeRegistry` + `NodeStakeUpdateValidation` |
| Modify | `LiveSequential.java` — same wiring |
| Create | `NodeStakeRegistryTest.java` — empty registry, update, dedup, time-lookup, persistence |
| Create | `NodeStakeUpdateValidationTest.java` — metadata, block processing, save/load roundtrip |
| Modify | `SignatureValidationTest.java` — `@Nested` stake-weighted tests |

Fixes #2491

### CODE REVIEW

**Review by jasperpotts — COMMENTED** [review]
### PR #2501 Review: Stake-Weighted Signature Validation                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                                            
  Overall Assessment: Looks correct with a few observations                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                            
  The PR correctly implements stake-weighted consensus validation and the intermittent state persistence pattern is solid.                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                            
  ---                                                                                                                                                                                                                                                                                                                                                                       
  **Weight Calculations - CORRECT**
                                                                                                                                                                                                                                                                                                                                                                            
  The PR extracts NodeStake.stake() (protobuf field 5) as the consensus weight. This is confirmed correct by both reference implementations:
                                                                                                                                                                                                                                                                                                                                                                            
  - Consensus node (EndOfStakingPeriodUtils.java): The stake field IS the clamped effective stake (clamp(stakeRewarded + stakeNotRewarded, 0 if < minStake, maxStake)) which becomes the roster weight.                                                                                                                                                                     
  - Mirror node (NodeStakeUpdateTransactionHandler.java): Uses nodeStakeProto.getStake() directly as the consensus weight.                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                            
  **Threshold Formula - CORRECT**                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                            
  The PR's stake-weighted threshold:                                                                                                                                                                                                                                                                                                                                        
  threshold = (totalStake / 3) + ((totalStake % 3 == 0) ? 0 : 1);
                                                                 
  This is ceil(totalStake / 3), which is mathematically identical to:                                                                                                                                                                                                                                                                                                       
  - Consensus node (RosterTransitionWeights.java:180): (totalWeight + 2) / 3 — also ceil(totalWeight / 3)                                                                                                                                                                                                                                                                   
  - Mirror node (ConsensusValidatorImpl.java:97-99): BigDecimal.valueOf(totalStake).multiply(1/3).setScale(0, CEILING) — also ceil(totalStake / 3)                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                            
  The comparison uses >= (line: if (validatedStake >= threshold)), matching:                                                                                                                                                                                                                                                                                                
  - Mirror node: BigDecimal.valueOf(stake).compareTo(stakeRequiredForConsensus) >= 0 (line 89)                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                                                                            
  Equal-Weight Fallback - Minor inconsistency (pre-existing)                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                                            
  The equal-weight fallback uses (totalNodes / 3) + 1 which is floor(n/3) + 1 (strictly >1/3), while the stake-weighted mode uses ceil(s/3) (≥1/3). These differ for multiples of 3:                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                            
  ```
┌───────┬────────────────────────┬───────────────────────────┐                                                                                                                                                                                                                                                                                                            
  │ Nodes │ Equal-weight threshold │ Stake-weighted equivalent │
  ├───────┼────────────────────────┼───────────────────────────┤                                                                                                                                                                                                                                                                                                            
  │ 3     │ 2 (>1/3)               │ 1 (≥1/3)                  │
  ├───────┼────────────────────────┼───────────────────────────┤                                                                                                                                                                                                                                                                                                            
  │ 6     │ 3 (=1/2)               │ 2 (=1/3)                  │                                                                                                                                                                                                                                                                                                            
  ├───────┼────────────────────────┼───────────────────────────┤
  │ 9     │ 4 (>1/3)               │ 3 (=1/3)                  │                                                                                                                                                                                                                                                                                                            
  └───────┴────────────────────────┴───────────────────────────┘      
```                                                                                                                                                                                                                                                                                                      
   
  This is not introduced by this PR — it was the pre-existing formula. The mirror node's equal-weight fallback (stake=1 per node, totalStake=nodeCount, then ceil(n/3)) would produce the lower values. The PR's approach is more conservative in equal-weight mode, which is arguably safer for the pre-staking era.                                                       
                  
  **State Persistence & Restart - CORRECT**                                                                                                                                                                                                                                                                                                                                     
                  
  The pattern is sound:                                                                                                                                                                                                                                                                                                                                                     
                  
  1. NodeStakeUpdateValidation implements save(Path) / load(Path) correctly, mirroring AddressBookUpdateValidation                                                                                                                                                                                                                                                          
  2. ToWrappedBlocksCommand saves on: shutdown hook, monthly checkpoints (inherited), and normal completion
  3. NodeStakeRegistry uses protobuf JSON serialization (NodeStakeHistory.JSON.write/parse) for reliable round-trips                                                                                                                                                                                                                                                        
  4. The +1ns timestamp offset ensures stakes apply to blocks AFTER the update, matching the AddressBookUpdateValidation pattern                                                                                                                                                                                                                                            
  5. Deduplication prevents redundant snapshots when stakes don't change                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                            
  **Stake Lookup Key - CORRECT**                                                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                                            
  stakeMap.getOrDefault(sig.nodeId(), 0L) uses nodeId, which matches NodeStakeEntry.nodeId(). Nodes not in the stake map get weight 0 (e.g., newly added nodes before the next NodeStakeUpdate), which is the correct conservative behavior.                                                                                                                                
   
  **Thread Safety - CORRECT**                                                                                                                                                                                                                                                                                                                                                   
                  
  CopyOnWriteArrayList is appropriate — writes are rare (daily NodeStakeUpdate) while reads happen during parallel signature validation.                                                                                                                                                                                                                                    
                  
  ---                                                                                                                                                                                                                                                                                                                                                                       
  **Issues / Suggestions**
                                                                                                                                                                                                                                                                                                                                                                            
  1. Performance (minor): getStakeMapForBlock() does a linear scan. With ~1 snapshot per day over years, this could grow to 1000+ entries. A binary search would be O(log n) instead of O(n). Not a correctness issue, but worth noting for long-running validations.
  2. Missing nodes in stake map: When a node has a valid RSA signature but is not in the stake map (e.g., new node added after the last NodeStakeUpdate), it gets weight 0. This is correct and conservative — the mirror node handles this the same way.                                                                                                                   
  3. Zero-stake nodes: Nodes with stake=0 in the NodeStakeUpdate are included in the map but contribute 0 to validated weight and 0 to total. The total stake still correctly reflects only non-zero-weight nodes' contributions since those with stake=0 add nothing. This matches the consensus node behavior where below-minStake nodes get weight 0.                    
  4. totalStake overflow: With long arithmetic, totalStake could theoretically overflow for extreme values, but in practice total stake on Hiero is well within long range (max ~50 billion HBAR = ~5 trillion tinybar, far below Long.MAX_VALUE).                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                            
  **Verdict**                                                                                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                                            
  The weight calculations and threshold logic are correct and consistent with both the consensus node and mirror node reference implementations. The state persistence pattern for restart/resume is well-implemented. The only asymmetry (equal-weight vs stake-weighted threshold formula) is pre-existing and the PR doesn't make it worse. This is safe to merge.

**Review by jsync-swirlds — APPROVED** [review]
Have we considered adding consensus-weighted validation, or have we decided stake-weight is close enough?

**Review by Nana-EC — APPROVED** [review]

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/SignatureValidation.java [hunk 1]
```diff
@@ -147,11 +165,31 @@ private void validateSignedRecordFileProof(
...
+        final long rawTotalStake = stakeMap != null
+                ? stakeMap.values().stream().mapToLong(Long::longValue).sum()
+                : 0;
+        final boolean stakeWeighted = stakeMap != null && rawTotalStake > 0;
+        final long totalStake;
+        final long threshold;
+        if (stakeWeighted) {
+            totalStake = rawTotalStake;
+            // Strong minority: ceil(totalStake / 3)
+            threshold = (totalStake / 3) + ((totalStake % 3 == 0) ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/SignatureValidation.java]: I think for now stake weight is correct for what we're doing (RSA signature validation on historical blocks). Consensus weight would only matter when we add TSS validation support and at that point we'd need to source weights from the roster rather than `NodeStakeUpdate` transactions anyway. I think we would need to add consensus weight when TSS support is added.

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/AddressBookUpdateValidation.java [hunk 2]
```diff
@@ -32,6 +32,8 @@
 public final class AddressBookUpdateValidation implements BlockValidation {
 
     private static final String SAVE_FILE_NAME = "addressBookHistory.json";
+    private static final int MAX_DEPTH = 512;
+    private static final int MAX_RECORD_FILE_SIZE = 64 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/validation/AddressBookUpdateValidation.java]: I will do in a follow up pr.
